sklearn/gaussian_process/gpc.py

```diff
@@ -0,0 +1,699 @@
+
+        if eval_gradient:
+            K, K_gradient = kernel(self.X_train_, eval_gradient=True)
+        else:
+            K = kernel(self.X_train_)
+
+        # Compute log-marginal-likelihood Z and also store some temporaries
+        # which can be reused for computing Z's gradient
+        Z, (pi, W_sr, L, b, a) = \
+            self._posterior_mode(K, return_temporaries=True)
+
+        if not eval_gradient:
+            return Z
+
+        # Compute gradient based on Algorithm 5.1 of GPML
+        d_Z = np.empty(theta.shape[0])
+        # XXX: Get rid of the np.diag() in the next line
+        R = W_sr[:, np.newaxis] * cho_solve((L, True), np.diag(W_sr))  # Line 7
+        C = solve(L, W_sr[:, np.newaxis] * K)  # Line 8
+        # Line 9: (use einsum to compute np.diag(C.T.dot(C))))
+        s_2 = -0.5*(np.diag(K) - np.einsum('ij, ij -> j', C, C)) \
+            * (pi * (1 - pi) * (1 - 2*pi))  # third derivative
+
+        for j in range(d_Z.shape[0]):
+            C = K_gradient[:, :, j]   # Line 11
+            # Line 12: (R.T.ravel().dot(C.ravel()) = np.trace(R.dot(C)))
+            s_1 = .5 * a.T.dot(C).dot(a) - .5 * R.T.ravel().dot(C.ravel())
+
+            b = C.dot(self.y_train_ - pi)  # Line 13
+            s_3 = b - K.dot(R.dot(b))  # Line 14
+
+            d_Z[j] = s_1 + s_2.T.dot(s_3)  # Line 15
+
+        return Z, d_Z
+
+    def _posterior_mode(self, K, return_temporaries=False):
+        """Mode-finding for binary Laplace GPC and fixed kernel.
+
+        This approximates the posterior of the latent function values for given
+        inputs and target observations with a Gaussian approximation and uses
+        Newton's iteration to find the mode of this approximation.
+        """
+        # Based on Algorithm 3.1 of GPML
+
+        # If warm_start are enabled, we reuse the last solution for the
+        # posterior mode as initialization; otherwise, we initialize with 0
+        if self.warm_start and hasattr(self, "f_cached") \
+           and self.f_cached.shape == self.y_train_.shape:
+            f = self.f_cached
+        else:
+            f = np.zeros_like(self.y_train_, dtype=np.float64)
+
+        # Use Newton's iteration method to find mode of Laplace approximation
+        log_marginal_likelihood = -np.inf
+        for _ in range(self.max_iter_predict):
+            # Line 4
+            pi = 1 / (1 + np.exp(-f))
+            W = pi * (1 - pi)
+            # Line 5
+            W_sr = np.sqrt(W)
+            W_sr_K = W_sr[:, np.newaxis] * K
+            B = np.eye(W.shape[0]) + W_sr_K * W_sr
+            L = cholesky(B, lower=True)
+            # Line 6
+            b = W * f + (self.y_train_ - pi)
+            # Line 7
+            a = b - W_sr * cho_solve((L, True), W_sr_K.dot(b))
+            # Line 8
+            f = K.dot(a)
+
+            # Line 10: Compute log marginal likelihood in loop and use as
+            #          convergence criterion
+            lml = -0.5*a.T.dot(f) \
```

Comment:
i think you're actually allowed to omit spaces around operators to indicate
precedence. but no pep8 checker gets that and, true, it isn't done that way
in sklearn

On Tuesday, August 25, 2015, Kyle Kastner <notifications@github.com> wrote:

> In sklearn/gaussian_process/gpc.py
> <https://github.com/scikit-learn/scikit-learn/pull/4270#discussion_r37865459>
> :
>
> > +            W = pi * (1 - pi)
> > +            # Line 5
> > +            W_sr = np.sqrt(W)
> > +            W_sr_K = W_sr[:, np.newaxis] * K
> > +            B = np.eye(W.shape[0]) + W_sr_K * W_sr
> > +            L = cholesky(B, lower=True)
> > +            # Line 6
> > +            b = W * f + (self.y_train_ - pi)
> > +            # Line 7
> > +            a = b - W_sr * cho_solve((L, True), W_sr_K.dot(b))
> > +            # Line 8
> > +            f = K.dot(a)
> > +
> > +            # Line 10: Compute log marginal likelihood in loop and use as
> > +            #          convergence criterion
> > +            lml = -0.5*a.T.dot(f) \
>
> Some pep8 strangeness here - should be spaces around operators like * but
> probably not between - and np.log
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4270/files#r37865459>.
>